clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsBuilderTest.java

```diff
@@ -219,6 +223,50 @@ public void testWriteEndTxnMarkerNonControlBatch() {
         builder.appendEndTxnMarker(RecordBatch.NO_TIMESTAMP, new EndTransactionMarker(ControlRecordType.ABORT, 0));
     }
 
+    @Test(expected = IllegalArgumentException.class)
+    public void testWriteLeaderChangeControlBatchWithoutLeaderEpoch() {
```

Comment:
I am not sure there is a need to be so strict about this. The `Log` class currently is responsible for adding the partition leader epoch. This is true even in our integration in `KafkaMetadataLog`. We use the following:
```scala
  override def appendAsLeader(records: Records, epoch: Int): lang.Long = {
    val appendInfo = log.appendAsLeader(records.asInstanceOf[MemoryRecords], leaderEpoch = epoch)
    appendInfo.firstOffset.getOrElse {
      throw new KafkaException("Append failed unexpectedly")
    }
  }
```
So whatever value we write in there on initialization is going to get overridden anyway.

Reply:
I see, still it makes no harm to push the check to early stage IMHO.